### PR TITLE
Add support for builtin executables

### PR DESCRIPTION
**Pull request by douglasjacobsen**
This merge adds the ability for application classes to define builtin executables, which are implemented as python methods.

These can be used in the list of executables for a workload. This work is in preparation for injecting arbitrary executables into the executables an experiment would use.

### CODE REVIEW

File: lib/ramble/ramble/application.py [hunk 1]
```diff
@@ -717,6 +701,53 @@ def _analysis_dicts(self, criteria_list):
                     }
         return files, contexts, foms
 
+    register_builtin('env_vars', required=True)
```
[the remark this answers is not in the document]
Comment (douglasjacobsen) [reply lib/ramble/ramble/application.py:704]: This is what makes it added to applications by default.

Applications can use this to omit it if they want, i.e.
```
register_builtin('env_vars', False)
```
In a sub-class would change it to not auto-inject.

Personally, I like it being a directive more than it being included in `__init__`. I feel like it's more annoying for a user to get the logic write (and a lot of application classes won't implement an `__init__` so, requiring one just to change the builtins that are required feels gross).

It's only currently the only magic builtin. I have several others I want to implement once the logic is decided upon, but adding those is tangential to this PR and would honestly just complicate this more.

File: var/ramble/repos/builtin.mock/applications/interleved-env-vars/application.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    executable('baz', 'foo', use_mpi=True)
+
+    input_file('input', url='file:///tmp/test_file.log',
+               description='Not a file', extension='.log')
+
+    workload('test_wl', executables=['builtin::env_vars', 'foo'], input='input')
+    workload('test_wl2', executables=['bar', 'builtin::env_vars'], input='input')
+    workload('test_wl3', executables=['baz'], input='input')
+
+    workload_variable('my_var', default='1.0',
```
[the remark this answers is not in the document]
Comment (douglasjacobsen) [reply var/ramble/repos/builtin.mock/applications/interleved-env-vars/application.py:26]: Sure, I can add one to this PR.

File: lib/ramble/ramble/language/application_language.py [hunk 3]
```diff
@@ -367,3 +367,35 @@ def _execute_success_criteria(app):
         }
 
     return _execute_success_criteria
+
+
+@application_directive('builtins')
+def register_builtin(name, required=False):
+    """Register a builtin method
+
+    Builtin methods are python methods that return lists of strings. These
```
[the remark this answers is not in the document]
Comment (douglasjacobsen) [reply lib/ramble/ramble/language/application_language.py:376]: Done

File: lib/ramble/ramble/application.py [hunk 4]
```diff
@@ -69,11 +75,23 @@ def __init__(self, file_path):
...
+        self._inject_required_builtins()
+
+    def _inject_required_builtins(self):
+        required_builtins = []
+        for builtin, blt_conf in self.builtins.items():
+            if blt_conf['required']:
+                required_builtins.append(builtin)
+
+        for workload, wl_conf in self.workloads.items():
+            if 'executables' in wl_conf:
```
[the remark this answers is not in the document]
Comment (douglasjacobsen) [reply lib/ramble/ramble/application.py:87]: Done

File: lib/ramble/ramble/application.py [hunk 5]
```diff
@@ -69,11 +75,23 @@ def __init__(self, file_path):
         self.application_class = 'ApplicationBase'
 
         self._verbosity = 'short'
+        self._inject_required_builtins()
+
+    def _inject_required_builtins(self):
+        required_builtins = []
+        for builtin, blt_conf in self.builtins.items():
+            if blt_conf['required']:
```
[the remark this answers is not in the document]
Comment (douglasjacobsen) [reply lib/ramble/ramble/application.py:83]: Done

File: lib/ramble/ramble/application.py [hunk 6]
```diff
@@ -47,9 +48,14 @@ def subsection_title(s):
     return level1_color + s + plain_format
 
 
+def nested_2(s):
```
[the remark this answers is not in the document]
Comment (douglasjacobsen) [reply lib/ramble/ramble/application.py:51]: Done

**Comment by douglasjacobsen** [comment]
@rfbgo Back to you. I force-pushed to fix the merge conflict. But changes do the the review are in the last commit in case you want to see them.

**Review by rfbgo — APPROVED** [review]
Thanks for the changes. LGTM

**Pull request merged**